Tangem/App/Services/Analytics/Analytics.swift

```diff
@@ -0,0 +1,243 @@
+//
+//  Analytics.swift
+//  Tangem
+//
+//  Created by Alexander Osokin on 31.03.2020.
+//  Copyright © 2020 Smart Cash AG. All rights reserved.
+//
+
+import Foundation
+import FirebaseAnalytics
+import FirebaseCrashlytics
+import AppsFlyerLib
+import BlockchainSdk
+import Amplitude
+import TangemSdk
+
+class Analytics {
+    @Injected(\.analyticsContext) private static var analyticsContext: AnalyticsContext
+
+    private init() {}
+
+    // MARK: - Scan
+
+    static func beginLoggingCardScan(source: CardScanSource) {
+        analyticsContext.set(value: source, forKey: .scanSource, scope: .common)
+        logInternal(source.cardScanButtonEvent)
+    }
+
+    static func endLoggingCardScan() {
+        guard let source = analyticsContext.value(forKey: .scanSource, scope: .common) as? CardScanSource else {
+            assertionFailure("Don't forget to call beginLoggingCardScan")
+            return
+        }
+
+        analyticsContext.removeValue(forKey: .scanSource, scope: .common)
+        logInternal(source.cardDidScanEvent)
+
+        if let cardId = analyticsContext.contextData?.analyticsParams[.cardId],
+           DemoUtil().isDemoCard(cardId: cardId) {
+            log(event: .demoActivated, params: [.cardId: cardId])
+        }
+    }
+
+    // MARK: - Others
+
+    static func logSignInIfNeeded(balance: Decimal) {
+        let isSignedIn = analyticsContext.value(forKey: .signedIn, scope: .userWallet) as? Bool ?? false
+        if isSignedIn {
+            return
+        }
+
+        analyticsContext.set(value: true, forKey: .signedIn, scope: .userWallet)
+
+        let params: [ParameterKey: String] = [
+            .state: ParameterValue.state(for: balance).rawValue,
+        ]
+
+        logInternal(.signedIn, params: params)
+    }
+
+    static func logTopUpIfNeeded(balance: Decimal) {
+        let previousBalance = analyticsContext.value(forKey: .balance, scope: .userWallet) as? Decimal
+
+        if balance > 0 {
+            if previousBalance == 0 {
+                logInternal(.toppedUp)
+                analyticsContext.set(value: balance, forKey: .balance, scope: .userWallet)
+            }
+        } else {
+            if previousBalance == nil { // do not save in a withdrawal case
+                analyticsContext.set(value: balance, forKey: .balance, scope: .userWallet)
+            }
+        }
+    }
```

Comment:
может собрать условия?

```suggestion
        if balance > 0, previousBalance == 0 {
                logInternal(.toppedUp)
                analyticsContext.set(value: balance, forKey: .balance, scope: .userWallet)
        } else if previousBalance == nil { // do not save in a withdrawal case
                analyticsContext.set(value: balance, forKey: .balance, scope: .userWallet)
         }
    }
```